### ISSUE TITLE
style+content transformer learning notes

### ISSUE THREAD

**Issue by WatershedXiaolan** [WatershedXiaolan/Xiaolan-s-Gitbook#110]
### What you need
1. A style image 
2. A content image
3. A pre-trained model to extract features
4. A ratio to decide the style vs content 

### Steps
1. Download the feature part of the pre-trained model, and freeze the parameters
2. Preprocessing content and style images so that they are of the same sizes and so on
3. Use the model to extract features for both style and content images
4. Apply Gram matrix to style features to extract style (Gram matrix shows correlations basically)
5. Initialize a target image; could be from content; enable the gradient calculation
6. Set up an optimizer and number of steps
7. for every step:
    - use the model to extract target feature
    - calculate content loss by a layer deep in the model `content_loss = torch.mean((target_features['conv4_2'] - content_features['conv4_2'])**2)`
    - calculate style loss by losses on multiple Gram matrixs `layer_style_loss = style_weights[layer] * torch.mean((target_gram - style_gram)**2)`
    - mix up the two losses by a ratio
    - loss backward and optimizer step
   

### Reference:
1. deep dream: https://deepdreamgenerator.com/
2. https://github.com/udacity/deep-learning-v2-pytorch
